In [ ]:
# default_exp core

# OHLCV Preprocessing

> This module holds all the functions necessary to preprocess the ohlcv data in a dataframe. The goal is to calculate what the highest percentage change was over the following n days. This will serve as a prediction target for ML purposes.

In [ ]:
#hide
from nbdev.showdoc import *

In [ ]:
import numpy as np
import pandas as pd
from pandas.io.json import json_normalize

In [ ]:
def json_parse(filename):
    data = pd.read_json('data/' + filename)
    data = json_normalize(data['result'])
    return data

In [ ]:
filename = 'btc-ltc-daily.json'

In [ ]:
data = json_parse(filename)

In [ ]:
data.head()

BV        C        H             L        O                    T  \
0  1.704125  0.02540  0.02540  2.525000e-02  0.02525  2014-03-07T00:00:00   
1  0.312342  0.02525  0.02525  2.525000e-02  0.02525  2014-03-10T00:00:00   
2  0.000339  0.02525  0.02525  2.525000e-02  0.02525  2014-03-11T00:00:00   
3  5.041498  0.02700  0.02700  2.540000e-02  0.02540  2014-03-12T00:00:00   
4  0.055185  0.01001  0.02525  1.000000e-08  0.02525  2014-03-13T00:00:00   

            V  
0   67.445876  
1   12.370000  
2    0.013434  
3  194.099661  
4    9.420337

In [ ]:
# truncate timestamp from T column (only call once, otherwise reload json into dataframe)
data['T'] = data['T'].str[:-9]

In [ ]:
data[0:30]

BV         C         H             L         O           T  \
0   1.704125  0.025400  0.025400  2.525000e-02  0.025250  2014-03-07   
1   0.312342  0.025250  0.025250  2.525000e-02  0.025250  2014-03-10   
2   0.000339  0.025250  0.025250  2.525000e-02  0.025250  2014-03-11   
3   5.041498  0.027000  0.027000  2.540000e-02  0.025400  2014-03-12   
4   0.055185  0.010010  0.025250  1.000000e-08  0.025250  2014-03-13   
5   0.181362  0.026700  0.026700  1.001000e-02  0.026700  2014-03-14   
6   0.193406  0.026700  0.026700  5.000000e-06  0.026700  2014-03-15   
7   3.074014  0.027900  0.270000  1.000000e-05  0.026700  2014-03-16   
8   1.744987  0.025800  0.028400  2.560000e-02  0.025700  2014-03-17   
9   1.325303  0.030300  0.035000  2.560000e-02  0.027900  2014-03-18   
10  0.216336  0.029000  0.031500  2.900000e-02  0.031500  2014-03-19   
11  0.105921  0.022600  0.022600  2.260000e-02  0.022600  2014-03-20   
12  0.054864  0.026000  0.026000  2.250000e-02  0.022500  2014-03-21   
13  0.777287  0.026000  0.029000  2.560001e-02  0.029000  2014-03-22   
14  0.109937  0.026300  0.030000  2.600000e-02  0.026000  2014-03-23   
15  0.058072  0.029400  0.030000  2.630000e-02  0.026300  2014-03-24   
16  0.210782  0.030000  0.030000  2.628000e-02  0.026280  2014-03-25   
17  0.506092  0.026756  0.029500  2.641000e-02  0.026600  2014-03-26   
18  0.458775  0.028856  0.028856  2.660500e-02  0.026756  2014-03-27   
19  0.219403  0.026666  0.028959  2.666000e-02  0.026700  2014-03-28   
20  0.316457  0.029290  0.029300  2.700000e-02  0.028000  2014-03-29   
21  0.422643  0.028995  0.029000  2.677021e-02  0.029000  2014-03-30   
22  0.580297  0.027800  0.029290  2.770100e-02  0.028995  2014-03-31   
23  0.771958  0.026605  0.027800  2.450000e-02  0.027700  2014-04-01   
24  0.533987  0.025650  0.027700  2.450000e-02  0.027000  2014-04-02   
25  0.342074  0.026970  0.027000  2.252000e-02  0.025650  2014-04-03   
26  0.539637  0.027700  0.027700  2.263600e-02  0.026970  2014-04-04   
27  0.159831  0.023333  0.025500  2.320001e-02  0.023200  2014-04-05   
28  0.105178  0.025200  0.026100  2.400000e-02  0.025500  2014-04-06   
29  0.074620  0.025720  0.027600  2.450090e-02  0.024501  2014-04-07   

             V  
0    67.445876  
1    12.370000  
2     0.013434  
3   194.099661  
4     9.420337  
5     6.926632  
6    11.862332  
7   177.266215  
8    64.057663  
9    47.424872  
10    7.201305  
11    4.686776  
12    2.147940  
13   29.579544  
14    4.098219  
15    1.968161  
16    7.753790  
17   18.803698  
18   16.955439  
19    8.070378  
20   11.113267  
21   15.301222  
22   20.781209  
23   28.077851  
24   20.699298  
25   14.306916  
26   21.733192  
27    6.828140  
28    4.229406  
29    2.929253

In [ ]:
data[0:30]['H'].max()
len(data.index)

2122

In [ ]:
highs = data.rolling(window=30, min_periods=0)['H'].max().shift(-30)
highs

0       0.27000
1       0.27000
2       0.27000
3       0.27000
4       0.27000
5       0.27000
6       0.27000
7       0.03500
8       0.03500
9       0.03150
10      0.03000
11      0.03000
12      0.03000
13      0.03000
14      0.03000
15      0.03000
16      0.02950
17      0.02930
18      0.02930
19      0.02930
20      0.02929
21      0.02929
22      0.02880
23      0.02880
24      0.02880
25      0.02880
26      0.02880
27      0.02880
28      0.02880
29      0.02880
         ...   
2092        NaN
2093        NaN
2094        NaN
2095        NaN
2096        NaN
2097        NaN
2098        NaN
2099        NaN
2100        NaN
2101        NaN
2102        NaN
2103        NaN
2104        NaN
2105        NaN
2106        NaN
2107        NaN
2108        NaN
2109        NaN
2110        NaN
2111        NaN
2112        NaN
2113        NaN
2114        NaN
2115        NaN
2116        NaN
2117        NaN
2118        NaN
2119        NaN
2120        NaN
2121        NaN
Name: H, Length: 2122, d

> Calculates the highest percentage change in the next n days compared to the close ('C').

In [ ]:
def high_pct_change(data, n):
    highs = data.rolling(window=n, min_periods=0)['H'].max().shift(-n)
    pcts = highs/data['C'] - 1
    return pcts

In [ ]:
pcts = high_pct_change(data, 30)
pcts.head()

0     9.629921
1     9.693069
2     9.693069
3     9.000000
4    25.973135
dtype: float64

In [ ]:
def round_pct(pct, interval=.25, cutoff=1):
    if pct >= cutoff:
        pct = cutoff
    elif pct < 0:
        pct = 0
    else:
        pct = interval*(pct // interval)
    return pct

In [ ]:
pct = round_pct(.45)
pct

0.25

> Takes list of percents and rounds them down based on interval. Anything above the cutoff is set to the cutoff. For example, if interval is 0.25 and cutoff is 1: 0.12 -> 0, 0.26 -> 0.25, 0.73 -> 0.50, 0.99 -> 0.75, 2.25 -> 1.00.

In [ ]:
def round_pcts(pcts, interval=.25, cutoff=1):
    pcts = pcts.apply(np.vectorize(round_pct))
    return pcts

In [ ]:
rounded_pcts = round_pcts(pcts)
rounded_pcts.head()

0    1.0
1    1.0
2    1.0
3    1.0
4    1.0
dtype: float64